wgsl/index.bs

```diff
@@ -6981,13 +6979,326 @@ See [[#statements]] and [[#function-calls]].
+  <tr><td class="nowrap">break;
+      <td>
+      <td>
+      <td>*CF*
+      <td>
+  <tr><td class="nowrap">continue;
+      <td>
+      <td>
+      <td>*CF*
+      <td>
+<tr><td class="nowrap">fallthrough;
+      <td>
+      <td>
+      <td>*CF*
+      <td>
+  <tr><td class="nowrap">discard;
+      <td>
+      <td>
+      <td>*CF*
+      <td>
+  <tr><td class="nowrap">return;
+      <td>
+      <td>
+      <td>*CF*
+      <td>*CF_return* -> *CF*
+  <tr><td class="nowrap">return *e*;
+      <td>
+      <td class="nowrap">(*CF*, *e*) => (*CF'*, *V*)
+      <td>*CF'*
+      <td>*CF_return* -> *CF'*<br>
+          *Value_return* -> *V*
+  <tr><td class="nowrap">e1 = e2;
+      <td>
+      <td class="nowrap">LValue: (*CF*, *e1*) => (*CF1*, *L1*)<br>
+          (*CF1*, *e2*) => (*CF2*, *V2*)
+      <td>*CF2*
+      <td>*L1* -> *V2*
+</table>
+
+### Uniformity rules for function calls ### {#uniformity-function-calls}
+
+The most complex rule is for function calls:
+- For each argument, apply the corresponding expression rule, with the control flow at the exit of the previous argument (using the control flow at the beginning of the function call for the first argument). Name the corresponding value nodes "arg_i" and the corresponding control flow nodes "CF_i"
+- Create two new nodes, named "Result" and "CF_after"
+- If the control-flow at the start of the function was tagged AlwaysMustBeUniform, then add an edge from MustBeUniform to the last CF_i
+- Otherwise add an edge from CF_after to the last CF_i
+- If the function was tagged ControlFlowCannotBeUniformAfter, then add an edge from CF_after to CannotBeUniform
+- Else if it was tagged ReturnValueCannotBeUniform, then add an edge from Result to CannotBeUniform
+- Add an edge from Result to CF_after
+- For each argument *i*:
+    - If the corresponding parameter was tagged AlwaysMustBeUniform, then add an edge from MustBeUniform to arg_i
+    - Else if it was tagged MustBeUniformForControlFlowAfter, then add an edge from CF_after to arg_i
+    - Else if it was tagged MustBeUniformForReturnValue, then add an edge from Result to arg_i
+
+Note: Notice that this rule only requires adding a number of edges bounded by 3 + the number of parameters of the functions, independently of how complex the implementation of the function might be. This is key to the linear complexity of the overall algorithm.
+
+Most built-in functions have tags of:
+- for the control-flow in which they can be called: MustBeUniformForControlFlowAfter
+- for the function as a whole: NoRestriction
+- for each parameter: MustBeUniformForReturnValue
+
+Here is the list of exceptions:
+- All functions in [[#derivative-builtin-functions]] are AlwaysMustBeUniform and ReturnValueCannotBeUniform
+- All functions in [[#texturesample]], [[#texturesamplebias]], [[#texturesamplecompare]], and [[#sync-builtin-functions]] are AlwaysMustBeUniform
+- TODO: what about [[#texturesamplelevel]] and [[#texturestore]] ?
```

Comment:
Both of these variants should just be NoRestriction. `textureSampleLevel` does not perform a derivative and `textureStore` is like an assignment (does not communicate with other invocations).

Reply:
Thanks for the explanation, I removed that TODO.